Iteration 7 Podcast Player

obenland opened the pull request:
Remaining work on Podcast Player Block

#### Changes proposed in this Pull Request:
* Use CSS mask to color ME.js controls (#15257)
* Introduce CSS vars for colors (#15321)
* Enable mediaelement player color customization (#15332)
* Simplify creating color change handler (#15335)
* Memoize colors object (#15336)
* Theme Compatibility Fixes (#15320)
* Lint everything (#15281)
* Fix Sound Icon Position (#15343)
* Remove bottom border on podcast header (#15355)
* Decode all HTML entities (#15351) 
* Consistent gutter spacing (#15356)
* Fix Sound Icon Position... including IE (#15371)
* Fix alignment issues in placeholder (#15376)
* Apply color customization to error state (#15394)
* Add an error boundary (#15369)
* Break components into their own files (#15442)
* Make speaker icons consistent (#15445) 
* Instruct IE11 user on how to activate audio support (#15451)
* Add external embed detection (#15446)
* Fix error message alignment (#15462)
* Make current track title inherit primary color (#15463)
* Center track icon vertically (#15465)

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Polishes Podcast Player block after CfT

#### Testing instructions:

_**NOTE: it should be tested in Simple sites**_

General URL for testing: https://anchor.fm/s/9400d7c/podcast/rss

* To test customization of controls colors (in modern browsers), go to the bottom of [`styles.css`](https://github.com/Automattic/jetpack/compare/feature/podcast-player...try/podcast-player-mejs-svg-colors?expand=1#diff-23936c23c59259d544403d0cc941ea1d) and set any color to the [background-color property of the ME.js button](https://github.com/Automattic/jetpack/compare/feature/podcast-player...try/podcast-player-mejs-svg-colors?expand=1#diff-23936c23c59259d544403d0cc941ea1dR340). Refresh the page and see if the color has been set. Play with the controls by clicking them - it should toggle colored `play`/`pause`, `mute`/`unmute` and `replay` SVGs.
* Play a track in the editor: The sound icon should look centered with the first line of the track
* Reduce the window size so the track titles wrap. The sound icon should still be centered with the first line of the track
* The sound icon being added should not increase the height of the track `<a>` element
* Test with one of these feeds and make sure there are no encoded html entities: https://whatsurfave.com/category/podcast/feed/, https://anchor.fm/s/9400d7c/podcast/rss
* Add the Podcast Player (Beta), Youtube embed, and RSS blocks. Compare the three blocks and check that input and button elements are aligned and that the blocks look consistent.
* Block a request / go offline to throw the episode error. In the editor, set primary/secondary/background colors and observe whether the error elements are now affected by color customization.
* Try adding `throw new Error( 'Test' )` somewhere in render functions of our components or inside functional components. For example, as the very first line in `render()` inside `audio-player.js`. Inside the editor, the player should be replaced by a simple error message. On the frontend, we have an additional handling that suppresses the error and instead unmounts the React component and brings back the static markup.
* In IE 11: Test both with the "audio support for webpages" setting unchecked and checked. Observe whether what you see reflects the states displayed in the #15451 issue description.
* Test with multiple URLs and confirm they all embed correctly. Some examples:
  - http://example.com - should show error
  - https://distributed.blog/category/podcast/feed/ - direct RSS link, should use our player
  - https://anchor.fm/startapodcast/episodes/Whats-your-podcast-about-e17krq/a-a2q3ft - regular webpage with `<link>` to RSS, should also use our player
  - https://open.spotify.com/show/7gozmLqbcbr6PScMjc0Zl4, invalid RSS but valid oEmbed, should replace Podcast Player with the spotify block
  - https://www.youtube.com/watch?v=dQw4w9WgXcQ, again invalid as RSS but works as oEmbed, should replace our block with youtube block


#### Proposed changelog entry for your changes:
* Not sure if needed at this time?

#### Follow-up issues
https://github.com/Automattic/jetpack/issues/15443, https://github.com/Automattic/jetpack/issues/15444, #15459

WunderBart committed:
Podcast Player: Use CSS mask to color ME.js controls (#15257)

Co-authored-by: Michael P. Pfeiffer <michael@cssconf.eu>

retrofox committed:
Podcast Player: introduce CSS vars for colors (#15321)

* add attributes to store the colors in HEX format

* store color in HEX

* podcast-player: add CSS variables in clint side

* add CSS vars in server side markup

* remove unused dependency

* Scope CSS variables to podcast player block

* sccope CSS variables, in server-side rendering

* improve doc, checking colors

* rename color handler

* add doc using multiline format

Co-authored-by: Michael P. Pfeiffer <michael@cssconf.eu>

frontdevde committed:
Enable mediaelement player color customization (#15332)

retrofox committed:
simplify creating color change handler (#15335)

marekhrabe committed:
Podcast Player: memoize colors object (#15336)

* extract colors object generator

* remove extraneous checks and optimizations

jeryj committed:
Podcast Player: Theme Compatibility Fixes (#15320)

* 2019 theme fixes
* Fixes compressed letter spacing in 2020 podcast title
* Increase specificity on header titles for improved consistency on frontend and editor in 2020
* Specify a current track podcast title color
* Fixes link underline while playing in the editor view on 2019 (and maybe others)

WunderBart committed:
Podcast Player: Lint everything (#15281)

* Reformat comments and lint all the files
* Make sure the "translators:" comments are being picked up
* Remove unused stuff in track-icons.js

jeryj committed:
Podcast Player: Fix Sound Icon Position (#15343)

* Uses absolute positioning on SVG instead of inline-block, which makes the wrapping span have an extra 4px on the height.

WunderBart committed:
Podcast Player: Fix accidental class instead of className

jeryj committed:
Remove bottom border on podcast header (#15355)

obenland committed:
Podcast Player: Decode all HTML entities (#15351)

* Decode all HTML entities

* Swap order of string manipulation

Props @marekhrabe

jeryj committed:
Podcast Player: Consistent gutter spacing (#15356)

* Fixes gutter spacing on tracklist and mejs to align with header padding

jeryj committed:
Podcast Player: Fix Sound Icon Position... including IE (#15371)

frontdevde committed:
Podcast player: fix alignment issues in placeholder (#15376)

frontdevde committed:
Podcast player: apply color customization to error state (#15394)

* Linting

* Podcast player: apply color customization to error state

* remove extraneous closing bracket

Co-authored-by: Marek Hrabe <marekhrabe@me.com>

marekhrabe committed:
Podcast Player: Add an error boundary (#15369)

* introduce error boundary and apply fallback HTML on frontend

* treat any audio errors before user interactions as critical failures

* typo

* reword the suggestion to reload

* Revert "treat any audio errors before user interactions as critical failures"

This reverts commit 8eaf71fa2a639c357205ccbc462a46b1f028f2f0.

marekhrabe commented on extensions/blocks/podcast-player/components/header.js:
```diff
@@ -9,12 +9,25 @@ import classnames from 'classnames';
 					<div className="jetpack-podcast-player__cover">
-						{ /* alt="" will prevent the src from being announced. Ideally we'd have a cover.alt, but we can't get that from the RSS */ }
+						{ /*
+						 * alt="" will prevent the src from being announced. Ideally we'd
```
should we mention "by a screen reader"?

marekhrabe committed:
Podcast Player: Break components into their own files (#15442)

* extract TrackIcon

* extract TrackError

* extract Track

frontdevde committed:
Podcast player: make speaker icons consistent (#15445)

* Podcast player: make speaker icons consistent

* Limit icon height to prevent sprite showing on multiline track titles

* Use icon size variable; use margin instead of mask offset

* Add explanation to hidden SVG

obenland committed:
[not verified] Clarify comment

Props @marekhrabe

frontdevde committed:
Podcast player: instruct IE11 user on how to activate audio support (#15451)

* Podcast player: instruct IE11 user on how to activate audio support

* Add descriptive error message and only show instructions on audio settings error

* Add notice that translation for error message is not needed

marekhrabe committed:
Podcast Player: Add external embed detection (#15446)

* initial implementation

* add proper embed checks, convert block to embed immediately

* remove createNotice, no longer used

* move api calls to an external function

WunderBart committed:
Podcast Player: Fix error message alignment (#15462)

WunderBart committed:
Podcast Player: Make current track title inherit primary color (#15463)

WunderBart committed:
Podcast Player: Center track icon vertically (#15465)

Making the line-height explicit prevents the icon from going off-middle in various themes

obenland commented:
@jeherve This PR should be ready to go. There's a video that demos our changes in paYJgx-Dh-p2, which might be a more comfortable way to review the functional changes.

I had to manually update the phablet diff due to the new files. Tests there are green.

Thanks again for your time & patience and taking on review of a PR of this size.

jeherve reviewed (approved):
This seems to work well. I left some minor comments below, but I don't think any of them are blockers, they can be addressed in a future PR.

******

I wonder if we could make the podcast links open in a new tab in the fallback markup.

******

I really like how the block can now handle multiple types of links and will transform nicely. I was also happily surprised that it will even handle regular RSS feeds (not podcasts) when some of the posts include audio tracks!

******

I think I already mentioned this in the past, but since we strip HTML markup from podcast descriptions, it makes the description field pretty much unusable whenever your podcast has things like lists or line breaks. Here is an example:
![image](https://user-images.githubusercontent.com/426388/79596202-6f815d00-80e0-11ea-96d6-8504f8afb4e3.png)

http://feeds.feedburner.com/lerendezvousjeux

I think it would be nice if we could keep some of the markup to avoid that.

******

As we're getting closer to launch, do you have any plans to work on `save` in the next iteration, so we can have some basic content saved to post content just in case folks get rid of Jetpack, or when the post is viewed outside of the WordPress context? I was thinking about creating an issue about it, but I wasn't sure how you all manage your to-do items and how to add a complexity note.
